local-cloudbuild.py

```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env python
+#
+# Run the steps of cloudbuild.yaml locally, emulating Google Container
+# Builder functionality.  Does not actually push images to the
+# Container Registry.  Not all functionality is supported.
+#
+# Based on https://cloud.google.com/container-builder/docs/api/build-steps
+
+# System packages
+import argparse
+import getpass
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+
+# Third party packages
+import yaml
+
+
+# Types
+class CloudBuildError(Exception):
+    pass
+
+
+def main(argv):
+    """Main entrypoint for cli"""
+    parser = argparse.ArgumentParser(
+        description='Process cloudbuild.yaml locally to build Docker images')
+    parser.add_argument(
+        'cloudbuild', type=str, help='Path to cloudbuild.yaml input file')
+    parser.add_argument(
+        '--keep_workspace',
+        type=bool,
+        default=False,
+        help='Retain workspace directory after building')
+    args = parser.parse_args(argv[1:])
+
+    # Load and parse cloudbuild.yaml
+    cloudbuild = None
+    with open(args.cloudbuild, 'rb') as infile:
+        cloudbuild = yaml.safe_load(infile)
+
+    host_workspace_parent = tempfile.mkdtemp(prefix='local-cloudbuild_%s_' %
+                                             getpass.getuser())
```

Comment:
Not need to namespace this by user, mkdtemp ensures its only readable by the current user.